[wasm] Add support for per-project customization of helix work items

radical opened the pull request:
# Motivation

The immediate motivation for this is the desire to run the `System.Security.Cryptography` tests - with, and without subtlecrypto. And only difference between the two runs is the headers that get set by passing `--web-server-use-cop` to `xharness`.

# Background
Currently, for sending tests to helix:
1. the project binaries are zipped, as part of the project build
2. then separate helix targets files build that adds helix items for
   those zip files.
    - for wasm, we test with multiple 'scenarios' - like v8, browser, and
    nodejs.
    - so, 3 helix work items are submitted per zip file

# Problem
- If a test project needs to have any customizations, for example, for
  testing crypto with a managed implementation, and subtlecrypto, which
  would require:
  - passing different arguments to xharness, in case of managed, and
    subtlecrypto
  - and this should be done only for the browser case
  - Currently, this would need this would need to be done in
  `sendtohelix-wasm.targets` special casing for the test project, and
  scenario.

# Solution

- Add support for running per-project extension targets
- To run a custom project specific target for adding/editing `@(HelixWorkItem)`:
- In the project add:
```xml
    <HelixTargetsFile Condition="'$(TargetOS)' == 'Browser'">wasm.helix.targets</HelixTargetsFile>
```
    - This file gets automatically copied next to the test archive as `$(MSBuildProjectName).helix.targets`

- In this `wasm.helix.targets` file, add the custom target to `$(HelixExtensionTargets)`

```xml
  <PropertyGroup Condition="'$(IsRunningLibraryTests)' == 'true' and '$(Scenario)' == 'WasmTestOnBrowser'">
    <HelixExtensionTargets>$(HelixExtensionTargets);_AddHelixCrypoItems</HelixExtensionTargets>
```

- The extension target will be called after the default items are added
  to `@(HelixWorkItem)`.
- Useful properties to condition on: `$(Scenario)`, `$(IsRunningLibraryTests)`
- All the properties in `src/libraries/sendtohelix-wasm.targets` will be available here
- And add to, change, or remove from @(HelixWorkItem)

Example:
```xml
  <Target Name="_AddHelixCrypoItems">
    <ItemGroup>
      <!-- remove the existing item -->
      <HelixWorkItem Remove="@(HelixWorkItem)" Condition="'%(OriginalFileName)' == '$(_CryptoProjectName)'" />

        <!-- add two new ones - managed, and subtylecrypto -->
      <HelixWorkItem Include="$(WorkItemPrefix)managed-$(_CryptoProjectName)">
        <PayloadArchive>$(TestArchiveTestsDir)$(_CryptoProjectName).zip</PayloadArchive>
        <Command>$(HelixCommand)</Command>
        <Timeout>$(_workItemTimeout)</Timeout>
        <OriginalFileName>$(_CryptoProjectName)</OriginalFileName>
      </HelixWorkItem>

      <HelixWorkItem Include="$(WorkItemPrefix)subtlecrypto-$(_CryptoProjectName)">
        <PayloadArchive>$(TestArchiveTestsDir)$(_CryptoProjectName).zip</PayloadArchive>
        <Command>$(HelixCommand)</Command>
        <Timeout>$(_workItemTimeout)</Timeout>
        <OriginalFileName>$(_CryptoProjectName)</OriginalFileName>

        <PreCommands Condition="'$(OS)' == 'Windows_NT'">set &quot;WasmXHarnessArgs=%WasmXHarnessArgs% --web-server-use-cop&quot;</PreCommands>
        <PreCommands Condition="'$(OS)' != 'Windows_NT'">export &quot;WasmXHarnessArgs=$WasmXHarnessArgs --web-server-use-cop&quot;</PreCommands>
      </HelixWorkItem>

      <_CryptoHelixItem Include="@(HelixWorkItem)"
                        Condition="$([System.String]::new('%(HelixWorkItem.Identity)').EndsWith('-$(_CryptoProjectName)'))" />
    </ItemGroup>

    <Error Text="Something went wrong. Expected to have only two work items for $(_CryptoProjectName). But got @(_CryptoHelixItem)"
           Condition="@(_CryptoHelixItem->Count()) != 2" />
  </Target>
```
- Remember that all these properties, and targets get imported into the
  msbuild *global* namespace, so make sure to use unique names to avoid
  conflicts with other test projects.

Future work: this commit only enables it for wasm/library tests, but it should
be easy to extract it out, but needs some testing.

# Also

This uses the above to run `System.Security.Cryptography.Tests` in two configurations on browser - one to test with subtlecrypto, and the other to test the fallback.

radical committed:
disable non-wasm builds

radical committed:
[wasm] Add support for per-project customization of helix work items

Currently, for sending tests to helix:
1. the project binaries are zipped, as part of the project build
2. then separate helix targets files build that adds helix items for
   those zip files.
    - for wasm, we test with multiple 'scenarios' - like v8, browser, and
    nodejs.
    - so, 3 helix work items are submitted per zip file

- If a test project needs to have any customizations, for example, for
  testing crypto with a managed implementation, and subtlecrypto, which
  would require:
  - passing different arguments to xharness, in case of managed, and
    subtlecrypto
  - and this should be done only for the browser case
  - Currently, this would need this would need to be done in
  `sendtohelix-wasm.targets` special casing for the test project, and
  scenario.

- We add support for importing `$(ProjectName).helix.targets`, and
  calling a special target in that to add helix items, as needed.

- This targets file can be set in the test project like:
```xml
<HelixTargetsFile Condition="'$(TargetOS)' == 'Browser'">wasm.helix.targets</HelixTargetsFile>
```
  - it will get deployed next to the zip file, and picked up automatically

```xml
<Project>
  <PropertyGroup>
    <_CryptoProjectName>System.Security.Cryptography.Tests</_CryptoProjectName>
    <System_Security_Cryptography_Tests_TargetName Condition="'$(Scenario))' == 'WasmTestOnBrowser'">System_Security_Cryptography_Tests_Targ
et</System_Security_Cryptography_Tests_TargetName>
  </PropertyGroup>

  <Target Name="System_Security_Cryptography_Tests_Target">
    <ItemGroup>
      <HelixWorkItem Include="$(Scenario)-managed-$(_CryptoProjectName)">
        <PayloadArchive>$(TestArchiveTestsDir)$(_CryptoProjectName).zip</PayloadArchive>
        <Command>$(HelixCommand)</Command>
        <Timeout>$(_workItemTimeout)</Timeout>
      </HelixWorkItem>

      <HelixWorkItem Include="$(Scenario)-subtlecrypto-System.Security.Cryptography.Tests">
        <PayloadArchive>$(TestArchiveTestsDir)$(_CryptoProjectName).zip</PayloadArchive>
        <Command>$(HelixCommand)</Command>
        <Timeout>$(_workItemTimeout)</Timeout>

        <PreCommands Condition="'$(OS)' == 'Windows_NT'">set &quot;WasmXHarnessArgs=--web-server-use-cors&quot;</PreCommands>
        <PreCommands Condition="'$(OS)' != 'Windows_NT'">export &quot;WasmXHarnessArgs=--web-server-use-cors&quot;</PreCommands>
      </HelixWorkItem>
    </ItemGroup>
  </Target>
```

- The targets file *must* have these:
    - a property named like `System_Security_Cryptography_Tests_TargetName`, for a
      test project named `System.Security.Cryptography.Tests`
    - if this property is not set, then the default behavior of adding
      work items will run
    - The target should add any items it needs to to `@(HelixWorkItem)`.
      - Examples of adding these can be seen in `sendtohelix*` project files

- Remember that all these properties, and targets get imported into the
  msbuild *global* namespace, so make sure to use unique names to avoid
  conflicts with other test projects.

Future work: this commit only enables it for wasm/library tests, but it should
be easy to extract it out, but needs some testing.

radical committed:
[wasm] Helix: test with, and without subtle crypto

radical commented:
cc @eerhardt @akoeplinger @steveisok

radical commented:
/azp run runtime-wasm

eerhardt commented on src/libraries/System.Security.Cryptography/tests/wasm.helix.targets:
```diff
@@ -0,0 +1,25 @@
+        <Command>$(HelixCommand)</Command>
+        <Timeout>$(_workItemTimeout)</Timeout>
+
+        <PreCommands Condition="'$(OS)' == 'Windows_NT'">set &quot;WasmXHarnessArgs=--web-server-use-cors&quot;</PreCommands>
```
This used to be called `--web-server-use-cop`, but here the `p` turned into `rs`. Is that intentional?

radical replied:
oops

eerhardt commented on src/libraries/System.Security.Cryptography/tests/wasm.helix.targets:
```diff
@@ -0,0 +1,25 @@
+        <Command>$(HelixCommand)</Command>
+        <Timeout>$(_workItemTimeout)</Timeout>
+
+        <PreCommands Condition="'$(OS)' == 'Windows_NT'">set &quot;WasmXHarnessArgs=--web-server-use-cors&quot;</PreCommands>
```
Do we need to append to the current WasmXHarnessArgs? Will this either overwrite, or get overwritten?

radical replied:
hm do you mean if the project itself is setting an *environment variable* with this name?
As for the *property*, that gets expanded, and written to the run script, so it won't be affected.

eerhardt replied:
how does the run script work? doesn't it set the environmnet variable? For example:

`
PATH=/workspaces/runtime/artifacts/bin/chromedriver/chromedriver_linux64:$PATH XHARNESS_ARGS="-v diag --browser-arg=--no-sandbox --browser-path /workspaces/runtime/artifacts/bin/chrome/chrome-linux/chrome" ./RunTests.sh
`

If it does this, doesn't that overwrite the "set"/"export" here?

radical replied:
If you are running this locally, then these targets won't be used. In case of helix, such arguments would get expanded into the script, like:

```
$HARNESS_RUNNER wasm $XHARNESS_COMMAND
	--app=. --output-directory=$XHARNESS_OUT
	-s dotnet.js.symbols --symbol-patterns wasm-symbol-patterns.txt --symbolicator WasmSymbolicator.dll,Microsoft.WebAssembly.Internal.SymbolicatorWrapperForXHarness
	$XHARNESS_ARGS
	$WasmXHarnessArgs
	-- 
	$WasmXHarnessMonoArgs
	--run WasmTestRunner.dll System.Security.Cryptography.Tests.dll $WasmTestAppArgs -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing $RSP_FILE
```

And we can set the environment variable like here. It should be safe to append anyway though. I'll add that.

eerhardt commented on eng/pipelines/common/platform-matrix.yml:
```diff
@@ -29,332 +29,332 @@ parameters:
 jobs:
 
 # Linux arm
-- ${{ if or(containsValue(parameters.platforms, 'Linux_arm'), in(parameters.platformGroup, 'all', 'gcstress')) }}:
```
Did the encoding get messed up here? Or some whitespace?

radical replied:
Nah, I commented out all this so I can only the wasm builds, and sanity check the PR first, and reiterate quicker if there are any issues.

radical committed:
address review feedback

radical committed:
Address review feedback

radical commented:
/azp run runtime-wasm

radical committed:
Fix typo

radical committed:
Revert "disable non-wasm builds"

This reverts commit 7ef99e81f82200189dd3f61eeaf00d6ca4ced6d4.

radical commented:
I'll move the bits in the wasm targets currently, outside to make it usable by others.

radical committed:
Merge branch 'dotnet:main' into helix-custom-targets

radical commented:
/azp run runtime-wasm

eerhardt commented on src/libraries/System.Security.Cryptography/tests/wasm.helix.targets:
```diff
@@ -0,0 +1,25 @@
+        <Timeout>$(_workItemTimeout)</Timeout>
+      </HelixWorkItem>
+
+      <HelixWorkItem Include="$(Scenario)-subtlecrypto-System.Security.Cryptography.Tests">
```
```suggestion
      <HelixWorkItem Include="$(Scenario)-subtlecrypto-$(_CryptoProjectName)">
```

to match the above.

eerhardt commented on src/libraries/sendtohelix-wasm.targets:
```diff
@@ -194,7 +196,27 @@
+
+      <!-- construct the property name that will contain the target to call for adding
+           helix items -->
+      <_TestWithCustomTarget PropertyName="$([System.String]::new(%(TestName)).Replace('.', '_').Replace('-', '_'))_TargetName" />
```
I wonder if this should be more specific than just "TestName_TargetName". Maybe it should say `Helix` in it: ex. "TestName_HelixCustomTargetName".

eerhardt commented on src/libraries/sendtohelix-wasm.targets:
```diff
@@ -194,7 +196,27 @@
+           items. For example, it might want to add custom helix items only for a nodejs scenario -->
+      <_TestWithCustomTarget Remove="@(_TestWithCustomTarget)" Condition="'$(%(_TestWithCustomTarget.PropertyName))' == ''" />
     </ItemGroup>
+    <Message Text="_TestWithCustomTarget: %(_TestWithCustomTarget.Identity) prop: %(_TestWithCustomTarget.PropertyName) archivePath: %(_TestWithCustomTarget.ArchivePath) tname: %(_TestWithCustomTarget.TestName)" />
```
```suggestion
    <Message Text="_TestWithCustomTarget: %(_TestWithCustomTarget.Identity) prop: %(_TestWithCustomTarget.PropertyName) archivePath: %(_TestWithCustomTarget.ArchivePath) testName: %(_TestWithCustomTarget.TestName)" />
```

radical replied:
oops, debug message. I'll remove it.

radical committed:
Update src/libraries/System.Security.Cryptography/tests/wasm.helix.targets

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

radical committed:
Merge remote-tracking branch 'origin/main' into helix-custom-targets

radical committed:
Address review feedback

radical commented:
I have a follow up PR uses this for Wasm.Build.Tests . I'll add some documentation in that.

radical committed:
remove debug spew

radical commented:
/azp run runtime-wasm

radical commented:
WasmBuildtests failure is https://github.com/dotnet/runtime/issues/70675 .

radical committed:
Merge remote-tracking branch 'origin/main' into helix-custom-targets

radical committed:
Change the way helix extension targets are discovered.

The new approach:

To run a custom project specific target for adding/editing @(HelixWorkItem):

- In the project add:
    `<HelixTargetsFile Condition="'$(TargetOS)' == 'Browser'">wasm.helix.targets</HelixTargetsFile>`

    - This file gets copied next to the test archive as $(MSBuildProjectName).helix.targets

- In this `wasm.helix.targets` file, add to $(HelixExtensionTargets) to run your custom target

```xml
  <PropertyGroup Condition="'$(IsRunningLibraryTests)' == 'true' and '$(Scenario)' == 'WasmTestOnBrowser'">
    <HelixExtensionTargets>$(HelixExtensionTargets);_AddHelixCrypoItems</HelixExtensionTargets>
```

- The extension target will be called after the default items are added
  to `@(HelixWorkItem)`.
- Useful properties to condition on: $(Scenario), $(IsRunningLibraryTests)
- And add to, change, or remove from @(HelixWorkItem)

Example:
```xml
  <Target Name="_AddHelixCrypoItems">
    <ItemGroup>
      <!-- remove the existing item -->
      <HelixWorkItem Remove="@(HelixWorkItem)" Condition="'%(OriginalFileName)' == '$(_CryptoProjectName)'" />

        <!-- add two new ones - managed, and subtylecrypto -->
      <HelixWorkItem Include="$(WorkItemPrefix)managed-$(_CryptoProjectName)">
        <PayloadArchive>$(TestArchiveTestsDir)$(_CryptoProjectName).zip</PayloadArchive>
        <Command>$(HelixCommand)</Command>
        <Timeout>$(_workItemTimeout)</Timeout>
        <OriginalFileName>$(_CryptoProjectName)</OriginalFileName>
      </HelixWorkItem>

      <HelixWorkItem Include="$(WorkItemPrefix)subtlecrypto-$(_CryptoProjectName)">
        <PayloadArchive>$(TestArchiveTestsDir)$(_CryptoProjectName).zip</PayloadArchive>
        <Command>$(HelixCommand)</Command>
        <Timeout>$(_workItemTimeout)</Timeout>
        <OriginalFileName>$(_CryptoProjectName)</OriginalFileName>

        <PreCommands Condition="'$(OS)' == 'Windows_NT'">set &quot;WasmXHarnessArgs=%WasmXHarnessArgs% --web-server-use-cop&quot;</PreCommands>
        <PreCommands Condition="'$(OS)' != 'Windows_NT'">export &quot;WasmXHarnessArgs=$WasmXHarnessArgs --web-server-use-cop&quot;</PreCommands>
      </HelixWorkItem>

      <_CryptoHelixItem Include="@(HelixWorkItem)"
                        Condition="$([System.String]::new('%(HelixWorkItem.Identity)').EndsWith('-$(_CryptoProjectName)'))" />
    </ItemGroup>

    <Error Text="Something went wrong. Expected to have only two work items for $(_CryptoProjectName). But got @(_CryptoHelixItem)"
           Condition="@(_CryptoHelixItem->Count()) != 2" />
  </Target>
```

radical commented:
/azp run runtime-wasm

radical committed:
cleanup

radical committed:
Move WBT specific stuff into the target too. This will make it simpler to move it off into a targets file later

radical commented:
/azp run runtime-wasm

radical committed:
fix build

radical commented:
/azp run runtime-wasm

radical committed:
fix libtests

radical committed:
fix wbt

radical commented:
/azp run runtime-wasm

radical committed:
[wasm] Bump helix timeout to 90mins for debugger tests on windows

radical commented:
@akoeplinger updated

radical commented:
/azp run runtime-wasm

eerhardt commented:
Is there any chance this PR can be reviewed/merged soon? I'd like to get https://github.com/dotnet/runtime/pull/70745 merged by Tuesday. That way HMAC on WASM is complete for the next preview release.

radical commented:
Thank you for the review, @steveisok!
